examples/vue/persister/src/persister.ts

```diff
@@ -0,0 +1,128 @@
+import type {
+  QueryClient,
+  QueryFunctionContext,
+  QueryKey,
+  QueryState,
+} from '@tanstack/query-core'
+import { hashQueryKey } from '@tanstack/query-core'
+
+export type Promisable<T> = T | PromiseLike<T>
+
+export interface PersistedQuery {
+  buster: string
+  timestamp: number
+  queryHash: string
+  queryKey: QueryKey
+  state: QueryState
+}
+
+export type PersistRetryer = (props: {
+  persistedQuery: PersistedQuery
+  error: Error
+  errorCount: number
+}) => Promisable<PersistedQuery | undefined>
+
+export interface AsyncStorage {
+  getItem: (key: string) => Promise<string | undefined | null>
+  setItem: (key: string, value: string) => Promise<void>
+  removeItem: (key: string) => Promise<void>
+}
+
+export interface StoragePersisterOptions<QC extends QueryClient> {
+  // TODO: if we decide to move this to an API layer, we could make this work without passing queryClient
+  /**
+   * Query Client instance
+   */
+  queryClient: QC
+  /** The storage client used for setting and retrieving items from cache.
+   * For SSR pass in `undefined`.
+   */
+  storage: AsyncStorage | Storage | undefined | null
+  /**
+   * How to serialize the data to storage.
+   * @default `JSON.stringify`
+   */
+  serialize?: (client: PersistedQuery) => string
+  /**
+   * How to deserialize the data from storage.
+   * @default `JSON.parse`
+   */
+  deserialize?: (cachedString: string) => PersistedQuery
+  /**
+   * A unique string that can be used to forcefully invalidate existing caches,
+   * if they do not share the same buster string
+   */
+  buster?: string
+  /**
+   * The max-allowed age of the cache in milliseconds.
+   * If a persisted cache is found that is older than this
+   * time, it will be discarded
+   */
+  maxAge?: number
+}
+
+export function createPersister<T, QC extends QueryClient>(
+  queryFn: () => Promise<T>,
+  {
+    queryClient,
+    storage,
+    buster = '',
+    maxAge = 1000 * 60 * 60 * 24,
+    serialize = JSON.stringify,
+    deserialize = JSON.parse,
+  }: StoragePersisterOptions<QC>,
+) {
+  return async (context: QueryFunctionContext) => {
+    const queryHash = hashQueryKey(context.queryKey)
+    const queryState = queryClient.getQueryState(context.queryKey)
+
+    if (!queryState?.data && storage != null) {
+      const storedData = await storage.getItem(queryHash)
+      if (storedData) {
+        const persistedQuery = deserialize(storedData) as PersistedQuery
+
+        if (persistedQuery.timestamp) {
+          const expired = Date.now() - persistedQuery.timestamp > maxAge
```

Comment:
I think we would also need to check if we have data in the cache already that is newer than data that we have in storage, no? For example, if dat is put into the cache via `queryClient.setQueryData` in the meantime. This check is basically built into the `hydrate()` function, so maybe we should be using that one instead of calling `.build()` directly?

also - that data wouldn't be persisted, as we are strictly bound to the `queryFn` right now, so if the queryFn doesn't run, there's no persistence. This can also happen if you prefetch on the server, then put data into the cache via `hydrate`.

Reply:
Correct me if i'm wrong, but query subscription and in turn `queryFn` would not be run if we are `restoring`?

https://github.com/TanStack/query/blob/beta/packages/react-query/src/useBaseQuery.ts#L62-L64
```ts
const unsubscribe = isRestoring
  ? () => undefined
  : observer.subscribe(notifyManager.batchCalls(onStoreChange))
```


This means that when we are in this piece of logic it's been already restored and we have something in the cache.
So initial condition for this branch would prevent restoring from persister and skip to `queryFn`:
https://github.com/TanStack/query/pull/5125/files/9ed0d9a942151772b5ac0f07ca9da539bde55824#diff-2a40d1fa40a7ae31330acd1ba80cfde7a697273dd9ae0d159bcba7c91e5a3959R79
```ts
const queryState = queryClient.getQueryState(context.queryKey)

if (!queryState?.data && storage != null) {
//      ^ do not restore if we already have something
```

Reply:
> but query subscription and in turn queryFn would not be run if we are restoring?

yes, that's true.

> This means that when we are in this piece of logic it's been already restored and we have something in the cache.

okay yeah that's also correct. It means that restoring would only work on empty caches though?

I'm just wondering why we wouldn't use the `hydrate` utility that we have already, which does a check that is slightly similar: It will always restore if data is newer, so that after hydration, you'll have the most up-to-date data, no matter where it came from ..,.

Reply:
We are inside a `queryFn` here, so we can't set it directly in queryCache, cause it will not render it until `queryFn` resolves. That is why we resolve the Promise with data from the persister.

As for
> It means that restoring would only work on empty caches though?

It there any way that `persisted data` will be newer than something that got added to the cache after client was instantiated in browser?
I do not think there will be any way that could happen.

Additionally by skipping this logic when we have data in `queryCache` already, we would not read it on every `queryFn` run, which is a perf improvement.

Reply:
> It there any way that persisted data will be newer than something that got added to the cache after client was instantiated in browser?

only if you get it into the storage by other means, so yeah, I get your point, and also 👍 about not running it in every queryFn